core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -2603,11 +2603,17 @@ object Parser {
      */
     def mergeCharIn[A, P0 <: Parser0[A]](ps: List[P0]): List[P0] = {
       @annotation.tailrec
-      def loop(ps: List[P0], front: List[(Int, BitSetUtil.Tpe)], result: Chain[P0]): Chain[P0] = {
+      def loop(ps: List[P0], front: List[CharIn], result: Chain[P0]): Chain[P0] = {
         @inline
         def frontRes: Chain[P0] =
-          if (front.isEmpty) Chain.nil
-          else Chain.one(Parser.charIn(BitSetUtil.union(front)).asInstanceOf[P0])
+          front match {
+            case Nil => Chain.nil
+            case one :: Nil => Chain.one(one.asInstanceOf[P0])
+            case many =>
+              // we need to union
+              val minBs: List[(Int, BitSetUtil.Tpe)] = many.map { case CharIn(m, bs, _) => (m, bs) }
+              Chain.one(Parser.charIn(BitSetUtil.union(minBs)).asInstanceOf[P0])
```

Comment:
could we have a version of `union` with `Iterator` and avoid the copy in `minBs`?

Reply:
we certainly could. I wasn't sure it was worth it because we already iterate through the number of parsers regularly, but the characters *inside* the parsers (which we have to iterate when allocating a new CharIn to compute the ranges) could be more expensive, which was what I was targeting here.

Do you want me to address this before merging?

Reply:
We can address this as a follow-up for sure!